src/IECorePython/TaskSchedulerInit.cpp

```diff
@@ -0,0 +1,96 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2018, Image Engine Design Inc. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of Image Engine Design Inc nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "boost/python.hpp"
+
+#include "tbb/tbb.h"
+
+using namespace boost::python;
+
+namespace
+{
+
+// Wraps task_scheduler_init so it can be used as a python
+// context manager.
+class TaskSchedulerInitWrapper : public tbb::task_scheduler_init
+{
+
+public :
+
+	TaskSchedulerInitWrapper(int max_threads)
+			: tbb::task_scheduler_init(deferred), m_maxThreads(max_threads)
+	{
+		if (max_threads != automatic && max_threads <= 0)
+		{
+			PyErr_SetString(PyExc_ValueError, "max_threads must be either automatic or a positive integer");
+			throw_error_already_set();
+		}
+	}
+
+	void enter()
+	{
+		initialize(m_maxThreads);
+	}
+
+	bool exit(boost::python::object excType, boost::python::object excValue, boost::python::object excTraceBack)
+	{
+		terminate();
+		return false; // don't suppress exceptions
+	}
+
+private :
+
+	int m_maxThreads;
+
+};
+
+} // namespace
+
+
+namespace IECorePython
+{
+
+void bindTaskSchedulerInit()
+{
+	object tsi = class_<TaskSchedulerInitWrapper, boost::noncopyable>( "_tbb_task_scheduler_init", no_init )
```

Comment:
I think we called this "_tbb_task_scheduler_init" in the Gaffer binding (leading underscore to denote protection) because we only needed it one place and we didn't know if we wanted to consider it as part of the API. If it's gonna be in Cortex that kindof implies it should be public so Gaffer can use it, because we'll need to remove the Gaffer binding to avoid conflict. Should we rename this to something else? Any thoughts @andrewkaufman?

Reply:
I'm fine just calling it `tbb_task_scheduler_init`, but we could also add a tbb namespace I suppose... `IECorePython.tbb.task_scheduler_init`

Reply:
I think last time we decided against the tbb namespace because we had no intention of wrapping anything else. Just dropping the `_` will do for me...

Reply:
> ImportError: /home/travis/build/ImageEngine/cortex/python/IECore/_IECore.so: undefined symbol: _ZN12IECorePython21bindTaskSchedulerInitEv

Probably missed the export as nm says: 
```00000000006f5820 t _ZN12IECorePython21bindTaskSchedulerInitEv```

Reply:
*gah* - missing include means the symbol gets defined but not exported.

Reply:
Report unresolved symbols in object files ``` -z defs``` and what to do if an unresolved symbol is encountered: ``` --error-unresolved-symbols ``` or ```--warn-unresolved-symbols```